### ISSUE TITLE
Lint request: Safety comment on safe code

### ISSUE THREAD

**Issue by alice-i-cecile** [rust-lang/rust-clippy#6880]
### What it does

Warns when a SAFETY comment is left on a safe function, trait etc.

*What is the advantage of the recommended code over the original code*

Avoids users accidentally using unsafe code in an inappropriate fashion because it's marked safe.

### Drawbacks

Won't work on normal comments, see #4828.

### Example

```rust
/// Safety: This function must be used very carefully 
pub fn my_fn(){
}
```

Should be written as:

```rust
/// Safety: This function must be used very carefully 
pub unsafe fn my_fn(){
}
```

Here's an [example](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/archetype.rs#L471-L516) of where this slipped through into real code.

**Comment by giraffate** [comment]
Do you mean the [`missing_safety_doc`](https://rust-lang.github.io/rust-clippy/master/index.html#missing_safety_doc) lint?

**Comment by alice-i-cecile** [comment]
No: the reverse. A safety doc was left on code that was not marked unsafe, implying that there were UB implications to using it incorrectly.

This is dangerous because safe code should not cause UB, even indirectly by violating unsafe code's assumptions in complex ways.

**Comment by ojeda** [comment]
This is a good idea; in fact, possibly for both `missing_safety_doc` and `undocumented_unsafe_blocks`.

The description is unclear which one it is referring to, though (it says "`SAFETY` comment" and "unsafe code", but the examples seem to imply safety contracts for unsafe functions).

**Comment by alice-i-cecile** [comment]
> The description is unclear which one it is referring to, though (it says "SAFETY comment" and "unsafe code", but the examples seem to imply safety contracts for unsafe functions).

My apologies. I meant the latter, but I suspect both would be useful.

**Comment by ojeda** [comment]
> My apologies. I meant the latter, but I suspect both would be useful.

No need to apologize! And thanks for this idea and opening this issue. I have opened another one for the former, then: https://github.com/rust-lang/rust-clippy/issues/7954.
